name: Lint Python code with Ruff & mypy

on:
  push:
    branches:
      - main
    paths:
      - '**.py'
  pull_request:
    paths:
      - '**.py'

permissions:
  contents: read

jobs:
  lint:
    runs-on: ubuntu-latest
    strategy:
      fail-fast: false
      matrix:
        vs-versions: [67]
        python-version: ['3.12']

    steps:
      - uses: actions/checkout@v4

      - uses: qoomon/actions--context@v3
        id: context
      - name: Determine branch name
        id: branch-names
        run: echo "current_branch=${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}}" >> $GITHUB_OUTPUT

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v5
        with:
          python-version: ${{ matrix.python-version }}

      - name: Set up uv
        uses: astral-sh/setup-uv@v6
        with:
          enable-cache: true

      - name: Set up VapourSynth ${{ matrix.vs-versions }}
        uses: ./.github/actions/setup-vapoursynth
        with:
          vapoursynth-version: ${{ matrix.vs-versions }}

      - name: Install dependencies
        run: |
          uv sync --locked

      - name: Running ruff
        run: uv run ruff check .

      - name: Running mypy
        run: uv run mypy vstools vssource vskernels vsexprtools vsrgtools vsmasktools vsscale vsaa vsdenoise

      - name: Post webhook for failure
        if: failure()
        uses: tsickert/discord-webhook@v6.0.0
        with:
          webhook-url: ${{ secrets.WEBHOOK_URL }}
          embed-url: ${{ steps.context.outputs.job_log_url }}
          embed-title: "[${{ github.event.repository.name }}] ${{ steps.context.outputs.job }} failed on ${{ steps.branch-names.outputs.current_branch }}"
          username: "GitHub - ${{ github.repository }}"
          avatar-url: https://github.githubassets.com/favicons/favicon.png
